make Cilium's CNI conf the only one available

aanm opened the pull request:
The initial implementation of CNI in Kubelet was made by checking for a valid configuration file, *.conf, in the configuration directory, which has the default of /etc/cni/net.d. This behavior allowed Cilium to be safely upgraded while other CNI configuration files were available as Kubelet would keep retrying Cilium’s configuration file until the Cilium CNI binary was available, since this binary is removed and re-added for each upgrade.

Unfortunately, this behavior has changed and Kubelet now picks the first valid CNI configuration file with a valid CNI binary. In a machine where multiple CNI configuration files are available, this might cause issues during a Cilium upgrade where Kubelet will use a different CNI plugin, for application pods, making those application pods to never be managed by Cilium.

To avoid this issue this commit changes he file extension of existing non-Cilium CNI configuration files from *.conf and *.conflist to *.conf.cilium_bak and *.conflist.cilium_bak respectively. These files will be reverted when the user wants to explicitly uninstall Cilium by performing a cilium cleanup command in the host. Since Kubelet only checks for the presence of *.conf and *.conflist it will not consider any other CNI configuration when setting up the application pods. This new behavior can be reverted by passing `--backup-cni-conf=false` to the postStart exec command in Cilium's DaemonSet.

Tested locally:

```
$ minikube start --network-plugin=cni --memory=4096 && minikube ssh -- sudo mount bpffs -t bpf /sys/fs/bpf

docker@minikube:~$ ls -la /etc/cni/net.d/
total 48
drwxr-xr-x. 1 root root 4096 Nov 26 21:07 .
drwxr-xr-x. 1 root root 4096 Aug 25 22:55 ..
-rw-r--r--. 1 root root  438 Jul 29 11:12 100-crio-bridge.conf
-rw-r--r--. 1 root root   54 Jun 22 18:23 200-loopback.conf
-rw-r--r--. 1 root root  639 May 25  2020 87-podman-bridge.conflist

docker@minikube:~$ ls -la /etc/cni/net.d/
total 52
drwxr-xr-x. 1 root root 4096 Nov 26 21:07 .
drwxr-xr-x. 1 root root 4096 Aug 25 22:55 ..
-rw-r--r--. 1 root root   97 Nov 26 21:07 05-cilium.conf
-rw-r--r--. 1 root root  438 Jul 29 11:12 100-crio-bridge.conf.cilium_bak
-rw-r--r--. 1 root root   54 Jun 22 18:23 200-loopback.conf.cilium_bak
-rw-r--r--. 1 root root  639 May 25  2020 87-podman-bridge.conflist.cilium_bak

docker@minikube:~$ ls -la /etc/cni/net.d/
total 52
drwxr-xr-x. 1 root root 4096 Nov 26 21:07 .
drwxr-xr-x. 1 root root 4096 Aug 25 22:55 ..
-rw-r--r--. 1 root root   97 Nov 26 21:07 05-cilium.conf
-rw-r--r--. 1 root root  438 Jul 29 11:12 100-crio-bridge.conf.cilium_bak
-rw-r--r--. 1 root root   54 Jun 22 18:23 200-loopback.conf.cilium_bak
-rw-r--r--. 1 root root  639 May 25  2020 87-podman-bridge.conflist.cilium_bak

docker@minikube:~$ ls -la /etc/cni/net.d/
total 52
drwxr-xr-x. 1 root root 4096 Nov 26 21:07 .
drwxr-xr-x. 1 root root 4096 Aug 25 22:55 ..
-rw-r--r--. 1 root root  438 Jul 29 11:12 100-crio-bridge.conf
-rw-r--r--. 1 root root   54 Jun 22 18:23 200-loopback.conf
-rw-r--r--. 1 root root  639 May 25  2020 87-podman-bridge.conflist
```

```release-note
Make Cilium the only CNI configuration available in the host to avoid pods from being managed by other CNIs while performing Cilium upgrades.
```

Fixes https://github.com/cilium/cilium/issues/14128
Fixes https://github.com/cilium/cilium/issues/13152

ti-mo committed:
plugins/cilium-cni: clean up Cilium CNI configs on start/stop

The uninstall script cleans up CNI binaries, but not the CNI config(list)s
pointing to them. This causes kubelet to spam the following log output:

```
Error validating CNI config list {.. full cni config here ...}:
[failed to find plugin "cilium-cni" in path [/opt/cni/bin]]
```

This commit cleans up CNI configs containing 'cilium' in the host's CNI config
directory when the Cilium agent is stopped.

This cleanup was previously omitted to prevent Pods from being scheduled
by other CNI plugins during a Cilium upgrade (when the agent is down).
It turns out that, since we remove the CNI binary during agent shutdown,
the Cilium CNI plugin is simply skipped, and Pod creation continues using
the node/cloud's next working CNI plugin.

A similar generic removal step was also added during CNI installation to
prevent backing up old Cilium CNI configs that would later be incorrectly
restored by the `cilium cleanup` command.

Fixes https://github.com/cilium/cilium/issues/13152.

Signed-off-by: Timo Beckers <timo@isovalent.com>

aanm committed:
cmd/cleanup: remove 05-cilium.conf

The 05-cilium.conf file is the default file use for CNI configuration,
this file should also be removed as part of the cleanup sub-command.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
plugins/cilium-cni: make Cilium's CNI conf the only one available

The initial implementation of CNI in Kubelet was made by checking for a
valid configuration file, *.conf, in the configuration directory, which
has the default of /etc/cni/net.d. This behavior allowed Cilium to be
safely upgraded while other CNI configuration files were available as
Kubelet would keep retrying Cilium’s configuration file until the Cilium
CNI binary was available, since this binary is removed and re-added for
each upgrade.

Unfortunately, this behavior has changed and Kubelet now picks the first
valid CNI configuration file with a valid CNI binary. In a machine where
multiple CNI configuration files are available, this might cause issues
during a Cilium upgrade where Kubelet will use a different CNI plugin,
for application pods, making those application pods to never be managed
by Cilium.

To avoid this issue this commit changes he file extension of existing
non-Cilium CNI configuration files from *.conf and *.conflist to
*.conf.cilium_bak and *.conflist.cilium_bak respectively. These files
will be reverted when the user wants to explicitly uninstall Cilium by
performing a cilium cleanup command in the host. Since Kubelet only
checks for the presence of *.conf and *.conflist it will not consider
any other CNI configuration when setting up the application pods. This
new behavior can be reverted by passing `--backup-cni-conf=false`
to the postStart exec command in Cilium's DaemonSet.

Tested locally:

```
$ minikube start --network-plugin=cni --memory=4096 && minikube ssh -- sudo mount bpffs -t bpf /sys/fs/bpf

docker@minikube:~$ ls -la /etc/cni/net.d/
total 48
drwxr-xr-x. 1 root root 4096 Nov 26 21:07 .
drwxr-xr-x. 1 root root 4096 Aug 25 22:55 ..
-rw-r--r--. 1 root root  438 Jul 29 11:12 100-crio-bridge.conf
-rw-r--r--. 1 root root   54 Jun 22 18:23 200-loopback.conf
-rw-r--r--. 1 root root  639 May 25  2020 87-podman-bridge.conflist

docker@minikube:~$ ls -la /etc/cni/net.d/
total 52
drwxr-xr-x. 1 root root 4096 Nov 26 21:07 .
drwxr-xr-x. 1 root root 4096 Aug 25 22:55 ..
-rw-r--r--. 1 root root   97 Nov 26 21:07 05-cilium.conf
-rw-r--r--. 1 root root  438 Jul 29 11:12 100-crio-bridge.conf.cilium_bak
-rw-r--r--. 1 root root   54 Jun 22 18:23 200-loopback.conf.cilium_bak
-rw-r--r--. 1 root root  639 May 25  2020 87-podman-bridge.conflist.cilium_bak

docker@minikube:~$ ls -la /etc/cni/net.d/
total 52
drwxr-xr-x. 1 root root 4096 Nov 26 21:07 .
drwxr-xr-x. 1 root root 4096 Aug 25 22:55 ..
-rw-r--r--. 1 root root   97 Nov 26 21:07 05-cilium.conf
-rw-r--r--. 1 root root  438 Jul 29 11:12 100-crio-bridge.conf.cilium_bak
-rw-r--r--. 1 root root   54 Jun 22 18:23 200-loopback.conf.cilium_bak
-rw-r--r--. 1 root root  639 May 25  2020 87-podman-bridge.conflist.cilium_bak

docker@minikube:~$ ls -la /etc/cni/net.d/
total 52
drwxr-xr-x. 1 root root 4096 Nov 26 21:07 .
drwxr-xr-x. 1 root root 4096 Aug 25 22:55 ..
-rw-r--r--. 1 root root  438 Jul 29 11:12 100-crio-bridge.conf
-rw-r--r--. 1 root root   54 Jun 22 18:23 200-loopback.conf
-rw-r--r--. 1 root root  639 May 25  2020 87-podman-bridge.conflist
```

Signed-off-by: André Martins <andre@cilium.io>
Signed-off-by: Timo Beckers <timo@isovalent.com>

aanm commented on plugins/cilium-cni/cni-install.sh:
```diff
@@ -169,6 +169,15 @@ if [ ! -d $(dirname $CILIUM_CNI_CONF) ]; then
 fi
 
+# Make sure only Cilium is used to setup an application pod. See GH-14128.
+find "$(dirname ${CILIUM_CNI_CONF})" \
```
depth 1

ti-mo commented on cilium/cmd/cleanup.go:
```diff
@@ -217,7 +217,7 @@ func (c ciliumCleanup) whatWillBeRemoved() []string {
 	toBeRemoved = append(toBeRemoved, fmt.Sprintf("endpoint state in %s",
 		defaults.RuntimePath))
-	toBeRemoved = append(toBeRemoved, fmt.Sprintf("CNI configuration at %s, %s, %s",
+	toBeRemoved = append(toBeRemoved, fmt.Sprintf("CNI configuration at %s, %s, %s, %s",
```
This change should probably be applied in the previous commit instead.

christarazi commented on plugins/cilium-cni/cni-install.sh:
```diff
@@ -83,6 +83,23 @@ if [ "${CILIUM_CUSTOM_CNI_CONF}" = "true" ]; then
 	exit 0
 fi
 
+
```
Seems like this commit can just be squashed with the previous commit.

ti-mo replied:
Done!

christarazi reviewed (approved):
LGTM, just one minor fixup

aanm commented on plugins/cilium-cni/cni-install.sh:
```diff
@@ -83,6 +84,14 @@ if [ "${CILIUM_CUSTOM_CNI_CONF}" = "true" ]; then
 
+# Remove any active Cilium CNI configurations left over from previous installs
+# to make sure the one we're installing later will take effect.
+echo "Removing active Cilium CNI configurations from ${CNI_CONF_DIR}})..."
```
Between this line and line 205 (`mv "${CNI_CONF_NAME}" "${CILIUM_CNI_CONF}"`) kubelet will be unable to create a pod. I know it will retry but ideally we should do `mv "${CNI_CONF_NAME}" "${CILIUM_CNI_CONF}"` first and then perform the deletes of the all confs that are not `${CNI_CONF_NAME}`

ti-mo replied:
This was done semi-intentionally because:
- this time window should be minimal
- this is during agent prestart, so not sure what Cilium CNI's pod creation behaviour is when the agent is not running yet
- there might be a timeframe where there are multiple Cilium CNI configs in the directory during upgrades, e.g. `05-cilium-cni.conflist` and a newer `05-cilium.conflist`.

I don't see the points above as problematic, it just seems safer to me to put the CNI directory into a clean state and only then install a new config. It also slightly simplifies the code.

aanm replied:
Answered inline.

> This was done semi-intentionally because:
> 
> * this time window should be minimal

yeah I know, but it can happen.

> * this is during agent prestart, so not sure what Cilium CNI's pod creation behaviour is when the agent is not running yet

Cilium CNI is smart enough and it will retry to connect to the agent multiple times before returns a failure to kubelet.

> * there might be a timeframe where there are multiple Cilium CNI configs in the directory during upgrades, e.g. `05-cilium-cni.conflist` and a newer `05-cilium.conflist`.
> 

IMO that is fine because in both cases Cilium is the one managing the network for the pod. Even if the 'wrong' CNI configuration is picked, that configuration is from Cilium.

> I don't see the points above as problematic, it just seems safer to me to put the CNI directory into a clean state and only then install a new config. It also slightly simplifies the code.

My main concern is during Cilium agent restarts, if the cilium pod is restarted, there will be a point in time where kubelet will fail hard to create pods because there isn't a single CNI configuration available.

ti-mo replied:
> Even if the 'wrong' CNI configuration is picked, that configuration is from Cilium.

True, but one might be a CNI chaining config, and the other might be a native IPAM config.

> if the cilium pod is restarted, there will be a point in time where kubelet will fail hard to create pods because there isn't a single CNI configuration available

IMO failing hard early is preferred over successfully launching a pod in the wrong networking config. The latter could mean  a security vulnerability (although it's a stretch).

aanm reviewed (commented):
@ti-mo just one small comment here

ti-mo committed:
cmd/cleanup: add long description to cilium cleanup command

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
plugins/cilium-cni: cni-install - document CILIUM_CUSTOM_CNI_CONF purpose

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
install/kubernetes: add cni.exclusive Helm flag

This exposes auto-management of the /etc/cni/net.d directory to the end user
to allow Cilium to be used with CNI managers like CNI-Genie.

Also document this as a breaking change to be released in 1.10.

Signed-off-by: Timo Beckers <timo@isovalent.com>

aanm commented on cilium/cmd/cleanup.go:
```diff
@@ -320,14 +328,34 @@ func confirmCleanup() bool {
+// revertCNIBackup removes the ".cilium_bak" suffix from all files in cniPath,
+// reverting them back to their original filenames.
+func revertCNIBackup() error {
+	return filepath.Walk(cniPath,
```
Just a small note here, `filepath.Walk` walks all directories recursive, not only the "top-level" one. It is likely to not affect anything but just a FYI.

christarazi reviewed (approved):
:rocket:

joestringer commented on plugins/cilium-cni/cni-install.sh:
```diff
@@ -72,8 +79,42 @@ fi
+# Cilium's CNI plugin will remain active. This makes sure Pods are not
+# scheduled by another CNI when the Cilium agent is down during upgrades
+# or restarts. See GH-14128 and related issues for more context.
+if [ "${CNI_EXCLUSIVE}" != "false" ]; then
```
Was the intent here to force users to opt out of exclusive control over the CNI?

I noticed that if the user fails to update their YAML files during the upgrade to this version, `CNI_EXCLUSIVE` will be undefined, and therefore does not match `false` and automatically takes this path.

ti-mo replied:
> Was the intent here to force users to opt out of exclusive control over the CNI?

Correct, it's opt-out since it can cause pods to be scheduled by other CNIs during Cilium upgrade/downtime. This is a potential security issue since network policies won't be enforced with those other CNIs. This has been documented in the upgrade notes. We didn't feel like this kind of change was appropriate to backport given its magnitude.

Didn't know about the `upgrade-impact` label, thanks!
